sdk/core/azure-core/src/main/java/com/azure/core/implementation/logging/DefaultLogger.java

```diff
@@ -0,0 +1,376 @@
+        logFromFormat(LogLevel.INFORMATIONAL, format, args);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isWarnEnabled() {
+        String logLevelStr = Configuration.getGlobalConfiguration().get(AZURE_LOG_LEVEL);
+        LogLevel currentLogLevel = LogLevel.fromString(logLevelStr);
+        return LogLevel.WARNING.getLogLevel() >= currentLogLevel.getLogLevel();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void warn(String format, Object arg) {
+        logFromFormat(LogLevel.WARNING, format, arg);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param format The formattable message to log.
+     * @param args Arguments for the message. If an exception is being logged, the last argument should be the
+     *     {@link Throwable}.
+     */
+    @Override
+    public void warn(String format, Object... args) {
+        logFromFormat(LogLevel.WARNING, format, args);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isErrorEnabled() {
+        String logLevelStr = Configuration.getGlobalConfiguration().get(AZURE_LOG_LEVEL);
+        LogLevel currentLogLevel = LogLevel.fromString(logLevelStr);
+        return LogLevel.ERROR.getLogLevel() >= currentLogLevel.getLogLevel();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void error(String format, Object arg) {
+        logFromFormat(LogLevel.ERROR, format, arg);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param format The formattable message to log.
+     * @param args Arguments for the message. If an exception is being logged, the last argument should be the
+     *     {@link Throwable}.
+     */
+    @Override
+    public void error(String format, Object... args) {
+        logFromFormat(LogLevel.ERROR, format, args);
+    }
+
+    /**
+     * Format and write the message according to the {@code MESSAGE_TEMPLATE}.
+     *
+     * @param level The level to log.
+     * @param format The log message format.
+     * @param arguments a list of arbitrary arguments taken in by format.
+     */
+    private void logFromFormat(LogLevel level, String format, Object... arguments) {
+        FormattingTuple tp = MessageFormatter.arrayFormat(format, arguments);
+        log(level, tp.getMessage(), tp.getThrowable());
+    }
+
+    /**
+     * Format and write the message according to the {@code MESSAGE_TEMPLATE}.
+     *
+     * @param level log level
+     * @param message The message itself
+     * @param t The exception whose stack trace should be logged
+     */
+    private void log(LogLevel level, String message, Throwable t) {
+        String dateTime = getFormattedDate();
+        String threadName = Thread.currentThread().getName();
+        String levelName = level.name();
+        StringBuilder buf = new StringBuilder(32);
+        buf.append(String.format(MESSAGE_TEMPLATE, dateTime, threadName, levelName, classPath, message));
+        writeWithThrowable(buf, t);
```

Comment:
If there is a `Throwable` to write is there a newline after the message template? Looking at the message template there is no newline inserted between the general log message and the stack trace.

Reply:
Added %n in template.